lib/util/string_conv/string_conv.c

```diff
@@ -0,0 +1,156 @@
+/*
+ * Copyright (c) 2022 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <ctype.h>
+#include <string.h>
+#include <stdlib.h>
+#include <logging/log.h>
+
+#define LOG_MODULE_NAME string_conv
+LOG_MODULE_REGISTER(LOG_MODULE_NAME, CONFIG_STRING_CONV);
+
+int string_conv_str2long(char *str, int *val)
+{
+	int str_len = strlen(str);
+	int temp_val;
+	int idx = 0;
+
+	if (!strncmp(str, "-", 1)) {
+		idx = 1;
+	}
+
+	for (int i = idx; i < str_len; i++) {
+		if (isdigit((int)str[i])) {
+			continue;
+		}
+
+		LOG_ERR("\"%s\" is not recognized as a valid signed decimal input", str);
+		return -EINVAL;
+	}
+
+	temp_val = strtol(str, NULL, 10);
+
+	if ((temp_val <= LONG_MIN) || (temp_val >= LONG_MAX)) {
+		LOG_ERR("Passed input value \"%s\" is too small/large", str);
+		return -EINVAL;
+	}
+
+	*val = temp_val;
+	return 0;
+}
+
+int string_conv_str2ulong(char *str, uint64_t *val)
+{
+	int str_len = strlen(str);
+	uint64_t temp_val;
+
+	for (int i = 0; i < str_len; i++) {
+		if (isdigit((int)str[i])) {
+			continue;
+		}
+
+		LOG_ERR("\"%s\" is not recognized as a valid unsigned decimal input", str);
+		return -EINVAL;
+	}
+
+	temp_val = strtoul(str, NULL, 10);
+
+	if (temp_val >= ULONG_MAX) {
+		LOG_ERR("Passed input value \"%s\" is too large", str);
+		return -EINVAL;
+	}
+
+	*val = temp_val;
+	return 0;
+}
+
+int string_conv_str2dbl(const char *str, double *val)
+{
+	char deci_buf[12], frac_buf[11] = { 0 };
+	int decimal;
+	uint64_t frac;
+	double frac_dbl;
+	int idx = strcspn(str, ",.");
+
+	if (idx > (sizeof(deci_buf) - 1)) {
+		LOG_ERR("Passed input value \"%s\" is too small/large", str);
+		return -EINVAL;
+	}
+
+	strncpy(deci_buf, str, idx);
+	strncpy(frac_buf, (str + (idx + 1)), MIN(sizeof(frac_buf) - 1, strlen(str + (idx + 1))));
+
+	if (string_conv_str2long(deci_buf, &decimal)) {
+		return -EINVAL;
+	}
+
+	if (string_conv_str2ulong(frac_buf, &frac)) {
+		return -EINVAL;
+	}
+
+	frac_dbl = (double)frac;
+
+	for (int i = 0; i < strlen(frac_buf); i++) {
```

Comment:
This, in bad scenario, evaluates `strlen` every loop run. Compiler may never be sure whether `frac_buf` is modified in background.

Reply:
End there is a lot of parsing and counting the same things allover again; for example `frac_buf` is filled by string_conv_str2ulong` that basically checks how many digits are there before then running `str2ulong` that does the same thing again and then we get here where the `frac_buf` gets processed to count digits.